From 8b183ed01c937e665dd7900ee9a80d68fceed88c Mon Sep 17 00:00:00 2001
From: Haihua Hu <jared.hu@nxp.com>
Date: Wed, 20 Mar 2024 19:14:51 +0900
Subject: [PATCH 46/68] MMFMWK-8195 waylandsink: add internal dmabuf pool to
 support sw decode

Add internal dmabuf pool to support SW decode. Propose to upstream if
available.

Upstream-Status: Pending
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 ext/wayland/gstwaylandsink.c | 111 +++++++++++++++++++++++++++++++----
 ext/wayland/gstwaylandsink.h |   2 +-
 ext/wayland/meson.build      |   8 ++-
 3 files changed, 109 insertions(+), 12 deletions(-)

diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index 17fff47..f134222 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -51,6 +51,10 @@
 #include <gst/video/videooverlay.h>
 #include <linux/version.h>
 
+#ifdef HAVE_DMABUFHEAPS_ALLOCATOR
+#include <gst/allocators/gstdmabufheaps.h>
+#endif
+
 #include <xf86drm.h>
 #include <xf86drmMode.h>
 
@@ -78,6 +82,8 @@ enum
   PROP_LAST
 };
 
+#define ISALIGNED(a, b) (!(a & (b-1)))
+
 GST_DEBUG_CATEGORY (gstwayland_debug);
 #define GST_CAT_DEFAULT gstwayland_debug
 
@@ -728,6 +734,8 @@ static gboolean
 gst_wayland_update_pool (GstWaylandSink * self, GstAllocator * allocator)
 {
   gsize size = self->video_info.size;
+  gint w = GST_VIDEO_INFO_WIDTH (&self->video_info);
+  gint h = GST_VIDEO_INFO_HEIGHT (&self->video_info);
   GstStructure *config;
 
   /* Pools with outstanding buffer cannot be reconfigured, so we must use
@@ -743,6 +751,24 @@ gst_wayland_update_pool (GstWaylandSink * self, GstAllocator * allocator)
   gst_buffer_pool_config_set_params (config, self->caps, size, 2, 0);
   gst_buffer_pool_config_set_allocator (config, allocator, NULL);
 
+  if (!ISALIGNED (w, 16) || !ISALIGNED (h, 16)) {
+    GstVideoAlignment alignment;
+
+    memset (&alignment, 0, sizeof (GstVideoAlignment));
+    alignment.padding_right = GST_ROUND_UP_N (w, 16) - w;
+    alignment.padding_bottom = GST_ROUND_UP_N (h, 16) - h;
+
+    GST_DEBUG
+        ("align buffer pool, w(%d) h(%d), padding_right (%d), padding_bottom (%d)",
+        w, h, alignment.padding_right, alignment.padding_bottom);
+
+    gst_buffer_pool_config_add_option (config,
+        GST_BUFFER_POOL_OPTION_VIDEO_META);
+    gst_buffer_pool_config_add_option (config,
+         GST_BUFFER_POOL_OPTION_VIDEO_ALIGNMENT);
+    gst_buffer_pool_config_set_video_alignment (config, &alignment);
+  }
+
   if (!gst_buffer_pool_set_config (self->pool, config))
     return FALSE;
 
@@ -807,6 +833,35 @@ gst_wayland_activate_drm_dumb_pool (GstWaylandSink * self)
   return TRUE;
 }
 
+static gboolean
+gst_wayland_activate_dmaheaps_pool (GstWaylandSink * self)
+{
+#ifdef HAVE_DMABUFHEAPS_ALLOCATOR
+  GstAllocator *alloc = NULL;
+
+  if (self->pool && gst_buffer_pool_is_active (self->pool)) {
+    GstStructure *config = gst_buffer_pool_get_config (self->pool);
+    gboolean is_dmaheaps = FALSE;
+
+    if (gst_buffer_pool_config_get_allocator (config, &alloc, NULL) && alloc)
+      is_dmaheaps = GST_IS_DMABUFHEAPS_ALLOCATOR (alloc);
+
+    gst_structure_free (config);
+
+    if (is_dmaheaps)
+      return TRUE;
+  }
+
+  alloc = gst_dmabufheaps_allocator_obtain ();
+  gst_wayland_update_pool (self, alloc);
+  gst_object_unref (alloc);
+
+  return TRUE;
+#else
+  return FALSE;
+#endif
+}
+
 static gboolean
 gst_wayland_sink_set_caps (GstBaseSink * bsink, GstCaps * caps)
 {
@@ -833,7 +888,7 @@ gst_wayland_sink_set_caps (GstBaseSink * bsink, GstCaps * caps)
   }
 
   self->video_info_changed = TRUE;
-  self->skip_dumb_buffer_copy = FALSE;
+  self->skip_dma_buffer_copy = FALSE;
 
   /* free pooled buffer used with previous caps */
   if (self->pool) {
@@ -892,7 +947,7 @@ gst_wayland_sink_propose_allocation (GstBaseSink * bsink, GstQuery * query)
   GstCaps *caps;
   GstBufferPool *pool = NULL;
   gboolean need_pool;
-  GstAllocator *alloc;
+  GstAllocator *alloc = NULL;
   guint64 drm_modifier;
   GstVideoInfoDmaDrm drm_info;
   GstVideoInfo vinfo;
@@ -930,9 +985,39 @@ gst_wayland_sink_propose_allocation (GstBaseSink * bsink, GstQuery * query)
     GstStructure *config;
     pool = gst_wl_video_buffer_pool_new ();
     config = gst_buffer_pool_get_config (pool);
+
+    gint w = GST_VIDEO_INFO_WIDTH (&self->video_info);
+    gint h = GST_VIDEO_INFO_HEIGHT (&self->video_info);
+
+    if (gst_wl_display_check_format_for_dmabuf (self->display, &self->drm_info)) {
+#ifdef HAVE_DMABUFHEAPS_ALLOCATOR
+      alloc = gst_dmabufheaps_allocator_obtain ();
+#endif
+    } else {
+      alloc = gst_shm_allocator_get ();
+    }
+
+    if (!ISALIGNED (w, 16) || !ISALIGNED (h, 16)) {
+      GstVideoAlignment alignment;
+
+      memset (&alignment, 0, sizeof (GstVideoAlignment));
+      alignment.padding_right = GST_ROUND_UP_N (w, 16) - w;
+      alignment.padding_bottom = GST_ROUND_UP_N (h, 16) - h;
+
+      GST_DEBUG_OBJECT
+          (self, "align buffer pool, w(%d) h(%d), padding_right (%d), padding_bottom (%d)",
+          w, h, alignment.padding_right, alignment.padding_bottom);
+
+      gst_buffer_pool_config_add_option (config,
+          GST_BUFFER_POOL_OPTION_VIDEO_ALIGNMENT);
+      gst_buffer_pool_config_set_video_alignment (config, &alignment);
+    }
+
     gst_buffer_pool_config_set_params (config, caps, size, 2, 0);
     gst_buffer_pool_config_set_allocator (config,
-        gst_shm_allocator_get (), NULL);
+        alloc, NULL);
+    gst_buffer_pool_config_add_option (config,
+          GST_BUFFER_POOL_OPTION_VIDEO_META);
     gst_buffer_pool_set_config (pool, config);
   }
 
@@ -940,11 +1025,13 @@ gst_wayland_sink_propose_allocation (GstBaseSink * bsink, GstQuery * query)
   if (pool)
     g_object_unref (pool);
 
-  alloc = gst_shm_allocator_get ();
-  gst_query_add_allocation_param (query, alloc, NULL);
+  if (alloc) {
+    gst_object_ref_sink (alloc);
+    gst_query_add_allocation_param (query, alloc, NULL);
+    g_object_unref (alloc);
+  }
   gst_query_add_allocation_meta (query, GST_VIDEO_META_API_TYPE, NULL);
   gst_query_add_allocation_meta (query, GST_VIDEO_CROP_META_API_TYPE, NULL);
-  g_object_unref (alloc);
 
   return TRUE;
 }
@@ -1126,7 +1213,7 @@ gst_wayland_sink_show_frame (GstVideoSink * vsink, GstBuffer * buffer)
       wbuf = gst_wl_linux_dmabuf_construct_wl_buffer (buffer, self->display,
           &self->drm_info);
 
-    if (!wbuf && !self->skip_dumb_buffer_copy) {
+    if (!wbuf && !self->skip_dma_buffer_copy) {
       /* DMABuf did not work, let try and make this a dmabuf, it does not matter
        * if it was a SHM since the compositor needs to copy that anyway, and
        * offloading the compositor from a copy helps maintaining a smoother
@@ -1134,8 +1221,12 @@ gst_wayland_sink_show_frame (GstVideoSink * vsink, GstBuffer * buffer)
        */
       GstVideoFrame src, dst;
 
-      if (!gst_wayland_activate_drm_dumb_pool (self)) {
-        self->skip_dumb_buffer_copy = TRUE;
+      if (gst_wayland_activate_dmaheaps_pool (self)) {
+        GST_WARNING_OBJECT (self, "active dmaheaps pool");
+      } else if (gst_wayland_activate_drm_dumb_pool (self)) {
+        GST_WARNING_OBJECT (self, "active drm dump pool");
+      } else {
+        self->skip_dma_buffer_copy = TRUE;
         goto handle_shm;
       }
 
@@ -1153,7 +1244,7 @@ gst_wayland_sink_show_frame (GstVideoSink * vsink, GstBuffer * buffer)
         if (G_UNLIKELY (!wbuf)) {
           GST_WARNING_OBJECT (self, "failed to import DRM Dumb dmabuf");
           gst_clear_buffer (&to_render);
-          self->skip_dumb_buffer_copy = TRUE;
+          self->skip_dma_buffer_copy = TRUE;
           goto handle_shm;
         }
 
diff --git a/ext/wayland/gstwaylandsink.h b/ext/wayland/gstwaylandsink.h
index c893618..d0ebe47 100644
--- a/ext/wayland/gstwaylandsink.h
+++ b/ext/wayland/gstwaylandsink.h
@@ -74,7 +74,7 @@ struct _GstWaylandSink
   GstVideoOrientationMethod current_rotate_method;
 
   gchar *drm_device;
-  gboolean skip_dumb_buffer_copy;
+  gboolean skip_dma_buffer_copy;
 
   /* fps print support */
   guint64 frame_showed;
diff --git a/ext/wayland/meson.build b/ext/wayland/meson.build
index f0a501a..36c3968 100644
--- a/ext/wayland/meson.build
+++ b/ext/wayland/meson.build
@@ -2,11 +2,17 @@ wl_sources = [
     'gstwaylandsink.c'
 ]
 
+dmabufheapsallocator_flags = []
+
+if cc.has_header('linux/dma-heap.h') and cc.has_header('gst/allocators/gstdmabufheaps.h', dependencies: gstallocators_dep)
+    dmabufheapsallocator_flags += ['-DHAVE_DMABUFHEAPS_ALLOCATOR']
+endif
+
 if use_wayland
 
     gstwaylandsink = library('gstwaylandsink',
         wl_sources + protocols_files,
-        c_args : gst_plugins_bad_args + ['-DGST_USE_UNSTABLE_API'],
+        c_args : gst_plugins_bad_args + dmabufheapsallocator_flags + ['-DGST_USE_UNSTABLE_API'],
         include_directories : [configinc],
         dependencies : [gst_dep, gstvideo_dep, gstwayland_dep,
                         wl_client_dep, wl_protocol_dep, libdrm_dep],
-- 
2.43.0

